services/api/src/utils/messaging/utils.js

```diff
@@ -1,76 +1,321 @@
+
+  const img = generateHtml('img', {
+    src: url,
+    alt,
+    ...dimensions,
+  });
+
+  return new Handlebars.SafeString(`<p>${img}</p>`);
+});
+
+Handlebars.registerHelper('imageUrl', (arg, options) => {
+  const upload = resolveUpload(arg);
+
+  if (!upload) {
+    return '';
+  }
+
+  return getImageUrl(upload, options.hash);
+});
+
+Handlebars.registerHelper('link', (...args) => {
+  const params = resolveParams(args, 'text', 'url');
+  return generateHtml('a', {
+    ...params,
+    target: '_blank',
+  });
+});
+
+Handlebars.registerHelper('button', (...args) => {
+  const params = resolveParams(args, 'text', 'url');
+  return generateHtml('a', {
+    ...params,
+    class: 'button',
+    target: '_blank',
+  });
+});
+
+Handlebars.registerHelper('date', (arg) => {
+  return getDateTime(arg).formatDate();
+});
+
+Handlebars.registerHelper('time', (arg) => {
+  return getDateTime(arg).formatTime();
+});
+
+Handlebars.registerHelper('rtime', (arg) => {
+  return getDateTime(arg).relative();
+});
+
+Handlebars.registerHelper('number', (arg) => {
+  const { index } = arg.data;
+  if (index == null) {
+    return '';
+  }
+  return index + 1;
+});
+
+// Allow fallback to default values.
+Handlebars.registerHelper('helperMissing', (value) => {
+  return typeof value === 'string' ? value : '';
+});
+
+// Helper utils
+
+function resolveParams(args, ...names) {
+  const ordered = args.slice(0, -1);
+  const [options] = args.slice(-1);
+  let params = { ...options.hash };
+  ordered.forEach((value, i) => {
+    const name = names[i];
+    params[name] = value;
+  });
+
+  params = normalizeParams(params);
+
+  return params;
+}
+
+function normalizeParams(params) {
+  const result = {};
+  for (let [key, value] of Object.entries(params)) {
+    if (key === 'url') {
+      key = 'href';
+      value = normalizeUrl(value, result);
+    }
+    result[key] = value;
+  }
+  return result;
+}
+
+const PARAM_REG = /:([a-z]+)/gi;
+
+function normalizeUrl(str, params) {
+  if (str.startsWith('/')) {
+    str = `${ENV['APP_URL']}${str}`;
+  }
+
+  str = str.replace(PARAM_REG, (_, key) => {
+    const value = params[key];
+    // Need to delete the injected params or they
+    // will be passed on to the HTML element.
+    delete params[key];
+    return value;
+  });
+
+  return str;
+}
+
+function resolveUpload(arg) {
+  if (Array.isArray(arg)) {
+    return arg[0];
+  }
+  return arg;
+}
+
+function generateHtml(tag, props) {
+  const { text, ...rest } = props;
+  const attr = Object.entries(rest)
+    .map((entry) => {
+      const [key, value] = entry;
+      if (value) {
+        return [key, `"${value}"`].join('=');
+      }
+    })
+    .filter((a) => a)
+    .join(' ');
+
+  let html = `<${tag} ${attr}>`;
+  if (text) {
+    html += `${text}</${tag}>`;
+  }
+  return new Handlebars.SafeString(html);
+}
+
+function interpolate(str, params) {
+  if (!str) {
+    return '';
+  }
+
+  params = {
+    ...ENV,
```

Comment:
not a fan of this ...I think its too risky. I dont a `stripe private key` regardless of how limit the access might be.

Reply:
We do have a prefix `APP_` that could serve to signify public envs but I don't know if it's enough. Other examples of  ENVs I would want to expose:

```
MEMBERS_EMAIL
NOTIFICATIONS_URL
TWILIO_WEBHOOK_URL
TWILIO_FROM_NUMBER
```

this is just a preliminary list. Rather than something like putting the onus of SECRET or something to protect private ENVs (I don't like that either for the record) maybe something like:

1. Anything that begins with APP_
2. Anything that ends with _URL
3. Anything that ends with _EMAIL

I think that would get me 95% of the way there.

Reply:
Resolved: Add comment in readme about this